---
title: "R Notebook"
output: html_notebook
---


*** 

# Apply clustering method
ref: 
https://bioconductor.org/packages/release/data/experiment/vignettes/DuoClustering2018/inst/doc/run_clustering.html




## Applying a new clustering algorithm to a provided data set

The code below describes how we applied each of the included clustering methods to the data sets for our paper (Duo, Robinson, and Soneson 2018). 

The  apply_*()  functions, describing how the respective clustering methods were run, are available from the GitHub repository corresponding to the publication. 

In order to apply a new clustering algorithm to one of the data sets using the same framework, it is necessary to generate a function with the same format. The input arguments to this function should be: 

* a **SingleCellExperiment** object

* a named list of parameter values (can be empty, if no parameters are used for the method)

* the desired number of clusters (k).

The function should return a list with three elements:

* **st** - a vector with the timing information. Should have five elements, named **user.self**, **sys.self**, **user.child**, **sys.child** and **elapsed**. 

* **cluster** - a named vector of cluster assignments for all cells.

* **est_k** - the number of clusters estimated by the method (if available, otherwise NA).

If the method does not allow specification of the desired number of clusters, but has another parameter affecting the resolution, this can be accommodated as well

***
First, load the package and define the data set and clustering method to use (note that in order to apply a method named <method>, there has to be a function named apply_<method>(), with the above specifications, available in the workspace).

load the package 
```{r}
suppressPackageStartupMessages({
  library(scater)
  library(SC3)
  library(Seurat)
  library(ggplot2)
  library(SingleCellExperiment)
  library(DuoClustering2018)
  library(ggplot2)
})
```

No non-system installation of Python could be found.
Would you like to download and install Miniconda?
Miniconda is an open source environment management system for Python.
See https://docs.conda.io/en/latest/miniconda.html for more details.

Choose sce_full_Zhengmix8eq as example data 

```{r}
sce_zheng <- sce_full_Zhengmix8eq(metadata = F)
```
mat_zheng is a large matrix with cell id and gene expression info
```{r}
mat_zheng = counts(sce_zheng)
dim(mat_zheng)
```
there are 15716 cells and 3994 genes.
```{r}
write.csv(mat_zheng,file = "D:\\New Term\\scRNA_clustering\\RNAClustering\\Dataset\\Zhengmix8eq.csv")
```

To get the true label of cells:(what we use as ground truth)

```{r}
labels = colData(sce_zheng)$phenoid
celltypes = data.frame(cellid = colnames(mat_zheng), label = labels)
```

```{r}
write.csv(celltypes,file = "D:\\New Term\\scRNA_clustering\\RNAClustering\\Dataset\\Zhengmix8eq_label.csv")

```

## k-means clustering with SC3 package
ref: https://github.com/markrobinsonuzh/scRNAseq_clustering_comparison/blob/master/Rscripts/clustering/apply_SC3.R

### 1. sc3 prepare

feature_symbol is defined as number of gene over there:

```{r}
rowData(sce_zheng)$feature_symbol <- rownames(counts(sce_zheng))
```

This method also defines all parameters needed for clustering and stores them in the sc3 slot. The parameters have their own defaults but can be manually changed. For more information on the parameters please use ?sc3_prepare.

```{r}
sce_zheng_1 <- sc3_prepare(sce_zheng)
```

### 2. estimate k


estimate the optimal k value:
When the sce object is prepared for clustering, SC3 can also estimate the optimal number of clusters k in the dataset. SC3 utilizes the Tracy-Widom theory on random matrices to estimate k.
```{r}
est_k <- metadata(sc3_estimate_k(sce_zheng_1))$sc3$k_estimation
```

```{r}
est_k
```
###3. calculate distances

Now we are ready to perform the clustering itself. First SC3 calculates distances between the cells. Method sc3_calc_dists calculates the distances, creates and populates the following items of the sc3 slot:

distances - contains a list of distance matrices corresponding to Euclidean, Pearson and Spearman distances.

```{r}
sce_zheng_2 <- sc3_calc_dists(sce_zheng_1)

```

###4. transformation

Next the distance matrices are transformed using PCA and graph Laplacian. Method sc3_calc_transfs calculates transforamtions of the distance matrices contained in the distances item of the sc3 slot. It then creates and populates the following items of the sc3 slot:

transformations - contains a list of transformations of the distance matrices corresponding to PCA and graph Laplacian transformations.

It also removes the previously calculated distances item from the sc3 slot:




```{r}
sce_zheng_3 <- sc3_calc_transfs(sce_zheng_2)

```


###5. k-means

kmeans should then be performed on the transformed distance matrices contained in the transformations item of the sc3 slot. Method sc3_kmeans creates and populates the following items of the sc3 slot:

kmeans - contains a list of kmeans clusterings.
By default the nstart parameter passed to kmeans defined in sc3_prepare method, is set 1000 and written to kmeans_nstart item of the sc3 slot. If the number of cells in the dataset is more than 2,000, this parameter is set to 50. A user can also manually define this parameter by changing the value of the kmeans_nstart item of the sc3 slot.

```{r}
sce_zheng_4 <- sc3_kmeans(sce_zheng_3, ks = 14)
```

### 6. sc3_calc_consens

When calculating consensus for each value of k SC3 averages the clustering results of kmeans using a consensus approach. Method sc3_calc_consens calculates consensus matrices based on the clustering solutions contained in the kmeans item of the sc3 slot. It then creates and populates the following items of the sc3 slot:

consensus - for each value of k it contains: a consensus matrix, an hclust object, corresponding to hierarchical clustering of the consensus matrix and the Silhouette indeces of the clusters.

```{r}
sce_zheng_5 <- sc3_calc_consens(sce_zheng_4)

```

```{r}
silhouette <- metadata(sce_zheng_5)$sc3$consensus$"14"$silhouette

silhouette<- data.frame(1:nrow(silhouette), silhouette[,1], silhouette[,2], silhouette[,3])
names(silhouette)<- c("id","cluster", "neighbor", "sil_width")
silhouette <- silhouette[order(silhouette$cluster,-silhouette$sil_width),]

ggplot(silhouette,aes(x = id, y = sil_width))+  geom_bar(stat = "identity")+geom_hline(aes(yintercept = .6))

```

           
next step: visualization ...

Rtsne + ggplot

Internal evaluation matrics




